### DISCUSSION TITLE
What is the intuition behind filamat shader variants?

### DISCUSSION THREAD

**Discussion by EnviralDesign** [google/filament#5410] [Q&A] [answered]
Hi, I'm trying to understand the intuition and logic of how these many shader variants that matinfo shows me are used in Filament. 
For example, if say I enable shadows in my scene, or enable rigging, etc. does this cause a different variant to be loaded at run time for that material? Is this process normally invisible to the end user making use of the Filament releases?

I'm extracting the shaders for use in an application I'm developing, and I'm attempting to implement them in a minimally intrusive manner.. so understanding how these work, the differences between them would be very helpful!
```
GLSL shaders:
    #0   gl41   vs 0x00
    #1   gl41   fs 0x00
    #2   gl41   vs 0x01   DIR
    #3   gl41   fs 0x01   DIR
    #4   gl41   vs 0x02   DYN
    #5   gl41   fs 0x02   DYN
    #6   gl41   vs 0x03   DIR|DYN
    #7   gl41   fs 0x03   DIR|DYN
    #8   gl41   vs 0x05   DIR|SRE
    #9   gl41   fs 0x05   DIR|SRE
    #10  gl41   vs 0x06   DYN|SRE
    #11  gl41   fs 0x06   DYN|SRE
    #12  gl41   vs 0x07   DIR|DYN|SRE
    #13  gl41   fs 0x07   DIR|DYN|SRE
    #14  gl41   vs 0x08   SKN
    #15  gl41   vs 0x09   DIR|SKN
    #16  gl41   vs 0x0a   DYN|SKN
    #17  gl41   vs 0x0b   DIR|DYN|SKN
    #18  gl41   vs 0x0d   DIR|SRE|SKN
    #19  gl41   vs 0x0e   DYN|SRE|SKN
    #20  gl41   vs 0x0f   DIR|DYN|SRE|SKN
    #21  gl41   vs 0x10   DEP
    #22  gl41   fs 0x10   DEP
    #23  gl41   vs 0x18   SKN|DEP
    #24  gl41   fs 0x20   FOG
    #25  gl41   fs 0x21   DIR|FOG
    #26  gl41   fs 0x22   DYN|FOG
    #27  gl41   fs 0x23   DIR|DYN|FOG
    #28  gl41   fs 0x25   DIR|SRE|FOG
    #29  gl41   fs 0x26   DYN|SRE|FOG
    #30  gl41   fs 0x27   DIR|DYN|SRE|FOG
    #31  gl41   fs 0x30   DEP|FOG
    #32  gl41   fs 0x44   SRE|VSM
    #33  gl41   fs 0x45   DIR|SRE|VSM
    #34  gl41   fs 0x46   DYN|SRE|VSM
    #35  gl41   fs 0x47   DIR|DYN|SRE|VSM
    #36  gl41   vs 0x50   DEP|VSM
    #37  gl41   fs 0x50   DEP|VSM
    #38  gl41   vs 0x58   SKN|DEP|VSM
    #39  gl41   fs 0x65   DIR|SRE|FOG|VSM
    #40  gl41   fs 0x66   DYN|SRE|FOG|VSM
    #41  gl41   fs 0x67   DIR|DYN|SRE|FOG|VSM
```

I found this useful ascii chart in variant.h:
```
    // DIR: Directional Lighting
    // DYN: Dynamic Lighting
    // SRE: Shadow Receiver
    // SKN: Skinning
    // DEP: Depth only
    // FOG: Fog
    // PCK: Picking (depth variant only)
    // VSM: Variance shadow maps
    //
    //   X: either 1 or 0
    //                               1     0    0      0     1     0     1
    //                      +-----+-----+-----+-----+-----+-----+-----+-----+
    // Variant              |  0  | VSM | FOG | DEP | SKN | SRE | DYN | DIR |   128
    //                      +-----+-----+-----+-----+-----+-----+-----+-----+
    //                                    PCK
    //
    // Standard variants:
    //                      +-----+-----+-----+-----+-----+-----+-----+-----+
    //                      |  0  | VSM | FOG |  0  | SKN | SRE | DYN | DIR |    64 - 22 = 42
    //                      +-----+-----+-----+-----+-----+-----+-----+-----+
    //      Vertex shader            0     0     0     X     X     X     X
    //    Fragment shader            X     X     0     0     X     X     X
    //       Fragment SSR            1     0     0     0     1     0     0
    //           Reserved            1     1     0     X     1     0     0      [ -2]
    //           Reserved            0     X     0     X     1     0     0      [ -4]
    //           Reserved            1     X     0     X     0     X     X      [-16]
    //
    // Depth variants:
    //                      +-----+-----+-----+-----+-----+-----+-----+-----+
    //                      |  0  | VSM | PCK |  1  | SKN |  0  |  0  |  0  |   8 - 2 = 6
    //                      +-----+-----+-----+-----+-----+-----+-----+-----+
    //       Vertex depth            X     0     1     X     0     0     0
    //     Fragment depth            X     X     1     0     0     0     0
    //           Reserved            1     1     1     X     0     0     0     [ -2]
    //
    // 48 variants used, 80 reserved (128 - 46)
    //
    // note: a valid variant can be neither a valid vertex nor a valid fragment variant
    //       (e.g.: FOG|SKN variants), the proper bits are filtered appropriately,
    //       see filterVariantVertex(), filterVariantFragment().
```

However I still feel like I'm missing something here...

For example if I take this line  `#41  gl41   fs 0x67   DIR|DYN|SRE|FOG|VSM`
Am I correct to assume that this shader variant can handle directional lighting, dynamic lighting, it can receive shadows, can simulate fog, and has support for variance shadow maps BUT it cannot work on a skinned / rigged mesh at the same time? Also why would a fragment shader have support for skinning? Isn't that vertex territory only?

Another example `#10  gl41   vs 0x06   DYN|SRE` I'd think that vertex program has no need or awareness of shadow mapping which is a fragment task, but then why is there a variant with that included?

How do I know which vertex and fragment programs to use together?

I assume I'm missing some crucial pieces to the puzzle, thanks as always!

**Comment by romainguy** [comment, marked as answer]
> Hi, I'm trying to understand the intuition and logic of how these many shader variants that matinfo shows me are used in Filament.
> For example, if say I enable shadows in my scene, or enable rigging, etc. does this cause a different variant to be loaded at run time for that material? Is this process normally invisible to the end user making use of the Filament releases?

That's correct yes.

> I'm extracting the shaders for use in an application I'm developing, and I'm attempting to implement them in a minimally intrusive manner.. so understanding how these work, the differences between them would be very helpful!

Wouldn't it be easier to just use Filament? 

> Am I correct to assume that this shader variant can handle directional lighting, dynamic lighting, it can receive shadows, can simulate fog, and has support for variance shadow maps BUT it cannot work on a skinned / rigged mesh at the same time? Also why would a fragment shader have support for skinning? Isn't that vertex territory only?

That's correct, and the fragment shader does not need support for skinning. If you look at the list you'll see that no fragment shader has that variant bit.

> Another example #10 gl41 vs 0x06 DYN|SRE I'd think that vertex program has no need or awareness of shadow mapping which is a fragment task, but then why is there a variant with that included?

Because the vertex shader does precomputations to speed up the fragment shader.

**Comment by EnviralDesign** [comment]
Ok thanks for clarity! 

I guess with all that said, I could also approach this by taking the variants that contain the most features (_# 20 and # 41 in the list above_) and dynamically "toggle" some of those hard coded defines. Might take a little more work and trial/error to figure out which defines to toggle together etc, but it would make less variants to manage changes to on my end.

---

> Wouldn't it be easier to just use Filament?

It's a good question :) I'm implementing these shaders in TouchDesigner. Touch is a node based environment with heavy extensibility possible using python and glsl. C++ is harder to implement, but can be done through their plugin templates. 

TBH If I was better with C++ I'd have attempted to integrate the entire project as my first strategy, but putting that aside I have some concerns anyways about how much dynamic runtime flexibility I'd have between Touch <-> Filament. 

Touch excels at exposing data and allowing direct and easy manipulation of it through nodes with out needing to re-compile/re-upload etc. I can see how assets get loaded and setup when a Filament project initializes, but it's not clear how easy or efficient it would be to make runtime changes to many of the assets.

Anyhow when I have something share worthy I will absolutely post it here. Thanks for all the amazing work and speedy help here.
